src/ucs/sys/topo.c

```diff
@@ -19,9 +53,125 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
+
+    ucs_topo_initialized = 1;
+}
+
+void ucs_topo_cleanup()
+{
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+
+    if (!ucs_topo_initialized) return;
+
+    kh_foreach_value(&ucs_topo_bus_to_sys_dev_hash, sys_dev, {
+        ucs_debug("sys_device_t %u not released", sys_dev);
+    })
+
+    kh_destroy_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_bus_to_sys_dev_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_sys_dev_to_bus_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    ucs_topo_initialized = 0;
+}
+
+static int ucs_topo_compare_bus_id(const ucs_sys_bus_id_t *bus_id1, 
+                                   const ucs_sys_bus_id_t *bus_id2)
+{
+    if ((bus_id1->domain == bus_id2->domain) &&
+        (bus_id1->bus == bus_id2->bus) &&
+        (bus_id1->slot == bus_id2->slot) &&
+        (bus_id1->function == bus_id2->function)) {
+        return 1;
+    }
+
+    return 0;
+}
+
 ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                                             ucs_sys_device_t *sys_dev)
 {
+    khiter_t hash_it;
+    ucs_topo_kh_put_t kh_put_status;
+
+    *sys_dev = UINT_MAX;
+
+    if (ucs_topo_compare_bus_id(bus_id, &ucs_sys_bus_id_unknown)) {
+        ucs_debug("found unknown device index %u for bus id %ld",
+                  *sys_dev, ucs_topo_get_bus_id_bit_repr(bus_id));
+        return UCS_OK;
+    }
+    
+    ucs_debug("find device index for bus id %ld", ucs_topo_get_bus_id_bit_repr(bus_id));
+
+    ucs_recursive_spin_lock(&ucs_topo_bus_to_sys_dev_lock);
+    hash_it = kh_put(bus_to_sys_dev/*name*/, 
+                     &ucs_topo_bus_to_sys_dev_hash /*pointer to hashmap*/,
+                     ucs_topo_get_bus_id_bit_repr(bus_id)/*key*/,
+                     &kh_put_status);
+
+    switch (kh_put_status) {
+        case UCS_KH_PUT_KEY_PRESENT:
+            *sys_dev = kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it);
+            ucs_debug("bus id %ld exists. sys_dev = %u",
+                      ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
+            break;
+        case UCS_KH_PUT_BUCKET_EMPTY:
+            *sys_dev = ucs_topo_sys_dev_to_bus_lookup.count;
+            kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it) = *sys_dev;
+            ucs_debug("bus id %ld doesn't exist. sys_dev = %u",
+                      ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
+
+            /* because of grabbing lock while holding another lock all other
+             * instances must make sure that the opposite pattern is not
+             * attempted to avoid deadlock */
+            ucs_recursive_spin_lock(&ucs_topo_sys_dev_to_bus_lock);
```

Comment:
lock already taken in line 140

Reply:
That's a different lock `bus_to_sys_dev_lock`